### PR TITLE
Add version, status to operator

### PR DESCRIPTION
**Pull request by sallyom**
Add version and status reporting
@enj @mrogers950 

TODO:
bump deps to pull in latest library-go, will do in follow-up PR
w/out this bump, status initiates to Available: True

### CODE REVIEW

File: pkg/operator/operator.go [hunk 1]
```diff
@@ -84,16 +91,46 @@ func (c serviceCAOperator) Key() (metav1.Object, error) {
...
+	operatorConfigCopy := operatorConfig.DeepCopy()
 	// This is to push out deployments but does not handle deployment generation like it used to. It may need tweaking.
-	err := sync_v4_00_to_latest(c, operatorConfig)
-	return err
+	syncErr := sync_v4_00_to_latest(c, operatorConfigCopy)
+	if syncErr != nil {
+		c.setFailingStatus(operatorConfigCopy, "OperatorSyncLoopError", syncErr.Error())
+	} else {
+		// Set current version and available status
+		version := os.Getenv("RELEASE_VERSION")
```
Comment (derekwaynecarr) [inline pkg/operator/operator.go]: can you confirm that you know that all the pods in openshift-service-ca are all running the latest version at this point and no old pods remain?  that is basically what we are trying to ensure when you set the version.

File: manifests/0000_09_service-ca-operator_07_clusteroperator.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+apiVersion: config.openshift.io/v1
+kind: ClusterOperator
+metadata:
+  name: service-ca
+spec: {}
```
Comment (enj) [inline manifests/0000_09_service-ca-operator_07_clusteroperator.yaml]: nit: remove

File: pkg/apis/serviceca/install.go [hunk 3]
```diff
@@ -1,26 +0,0 @@
-package kubeapiserver
```
Comment (enj) [inline pkg/apis/serviceca/install.go]: separate delete commit

File: manifests/0000_09_service-ca-operator_05_deploy.yaml [hunk 4]
```diff
@@ -28,6 +28,8 @@ spec:
         env:
         - name: CONTROLLER_IMAGE
           value: quay.io/openshift/origin-service-ca-operator:v4.0
+        - name: RELEASE_VERSION
```
Comment (enj) [inline manifests/0000_09_service-ca-operator_05_deploy.yaml]: `OPERATOR_IMAGE_VERSION`

File: pkg/operator/sync_apiservice_v4_00.go [hunk 5]
```diff
@@ -63,7 +63,7 @@ func syncAPIServiceController_v4_00_to_latest(c serviceCAOperator, operatorConfi
 	}
 
 	// TODO create a new configmap whenever the data value changes
-	_, configMapModified, err := manageAPIServiceConfigMap_v4_00_to_latest(c.corev1Client, c.eventRecorder, operatorConfig)
+	_, configMapModified, err := manageAPIServiceConfigMap_v4_00_to_latest(c.corev1Client, c.eventRecorder)
```
Comment (enj) [inline pkg/operator/sync_apiservice_v4_00.go]: Keep passing this through and add TODO to combine CLI config and allow overrides with unsupported config overrides.

File: pkg/operator/sync_signer_v4_00.go [hunk 6]
```diff
@@ -66,7 +66,7 @@ func syncSigningController_v4_00_to_latest(c serviceCAOperator, operatorConfig *
 	}
 
 	// TODO create a new configmap whenever the data value changes
-	_, configMapModified, err := manageSigningConfigMap_v4_00_to_latest(c.corev1Client, c.eventRecorder, operatorConfig)
+	_, configMapModified, err := manageSigningConfigMap_v4_00_to_latest(c.corev1Client, c.eventRecorder)
```
Comment (enj) [inline pkg/operator/sync_signer_v4_00.go]: Same comment.

File: pkg/operator/operator.go (hunk 1 above)
Comment (sallyom) [reply pkg/operator/operator.go]: I've updated the status reporting, will test to be sure it has the correct behavior

File: manifests/0000_09_service-ca-operator_07_clusteroperator.yaml (hunk 2 above)
Comment (sallyom) [reply manifests/0000_09_service-ca-operator_07_clusteroperator.yaml]: done

File: pkg/apis/serviceca/install.go (hunk 3 above)
Comment (sallyom) [reply pkg/apis/serviceca/install.go]: done

File: manifests/0000_09_service-ca-operator_05_deploy.yaml (hunk 4 above)
Comment (sallyom) [reply manifests/0000_09_service-ca-operator_05_deploy.yaml]: done

File: manifests/0000_09_service-ca-operator_07_clusteroperator.yaml (hunk 2 above)
Comment (sallyom) [reply manifests/0000_09_service-ca-operator_07_clusteroperator.yaml]: removed

File: pkg/operator/sync_apiservice_v4_00.go (hunk 5 above)
Comment (sallyom) [reply pkg/operator/sync_apiservice_v4_00.go]: follow-up ok?

File: pkg/operator/sync_signer_v4_00.go (hunk 6 above)
Comment (sallyom) [reply pkg/operator/sync_signer_v4_00.go]: follow-up ok?

File: pkg/operator/operator.go [hunk 7]
```diff
@@ -85,15 +93,43 @@ func (c serviceCAOperator) Key() (metav1.Object, error) {
...
+		if setOperatorVersion {
+			version := os.Getenv("OPERATOR_IMAGE_VERSION")
+			if c.versionGetter.GetVersions()["operator"] != version {
+				glog.Infof("Updating clusteroperator %s version: %v", clusterOperatorName, version)
+				// Set current version
+				c.versionGetter.SetVersion("operator", version)
+			}
+		}
+	default:
+		glog.Warningf("unknown custom resource state: %s", operatorConfig.Spec.ManagementState)
```
Comment (stlaz) [inline pkg/operator/operator.go]: I think we shouldn't even try to check the statuses in this case.

File: pkg/operator/status.go [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+func (c *serviceCAOperator) syncStatus(operatorConfigCopy *operatorv1.ServiceCA, deployments []string) (bool, error) {
+	ready := 0
+	existingDeploymentsAndReplicas := 0
+	var statusErr error
+	for _, dep := range deployments {
+		existing, err := c.appsv1Client.Deployments(operatorclient.TargetNamespace).Get(dep, metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			glog.Infof("Deployment %s does not exist", dep)
+			statusMsg := fmt.Sprintf("Deployment %s does not exist", dep)
```
Comment (stlaz) [inline pkg/operator/status.go]: I don't think `statusMsg` should appear twice in the glog (once here, once when `Sync()` returns failure). Pick one.

File: pkg/operator/status.go [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+				ready++
+			}
+		}
+	}
+	// only report version if all deployments are available and updated
+	if ready == len(deployments) {
+		glog.Infof("All deployments managed by service-ca-operator are available and updated")
+		c.setAvailableStatus(operatorConfigCopy)
+		return true, statusErr
+		// set Available if replica instances are created.  Don't report new version if
```
Comment (stlaz) [inline pkg/operator/status.go]: This comment should be inside the condition branch it applies to.

File: pkg/operator/status.go [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+	existingDeploymentsAndReplicas := 0
+	var statusErr error
+	for _, dep := range deployments {
+		existing, err := c.appsv1Client.Deployments(operatorclient.TargetNamespace).Get(dep, metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			glog.Infof("Deployment %s does not exist", dep)
+			statusMsg := fmt.Sprintf("Deployment %s does not exist", dep)
+			c.setFailingStatus(operatorConfigCopy, "Error getting service-ca-operator managed deployments", statusMsg)
+			return false, statusErr
+		} else if existing.DeletionTimestamp != nil {
```
Comment (stlaz) [inline pkg/operator/status.go]: Each of these if's end with return, no need to do `else if`

File: pkg/operator/starter.go [hunk 11]
```diff
@@ -64,18 +75,28 @@ func RunOperator(ctx *controllercmd.ControllerContext) error {
...
+	versionGetter := status.NewVersionGetter()
+	clusterOperator, err := configClient.ConfigV1().ClusterOperators().Get(clusterOperatorName, metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	for _, version := range clusterOperator.Status.Versions {
+		versionGetter.SetVersion(version.Name, version.Version)
+	}
+	versionGetter.SetVersion("operator", os.Getenv("OPERATOR_IMAGE_VERSION"))
```
Comment (stlaz) [inline pkg/operator/starter.go]: I know I saw this code in a different repo, too, but won't this just flip the version of `operator` before we perform the checks that we should be updating it?

File: pkg/operator/operator.go (hunk 7 above)
Comment (sallyom) [reply pkg/operator/operator.go]: agreed, thanks, done

File: pkg/operator/status.go (hunk 8 above)
Comment (sallyom) [reply pkg/operator/status.go]: removed it here

File: pkg/operator/status.go (hunk 9 above)
Comment (sallyom) [reply pkg/operator/status.go]: fixed

File: pkg/operator/starter.go (hunk 11 above)
Comment (sallyom) [reply pkg/operator/starter.go]: yes, it's in cluster-kube-apiserver, pkg/operator/starter.go.  I agree.  I was confused, meant to ask about that, I'll ask them, leave it out for now.  (I added that yesterday late and forgot about it, thanks)

File: pkg/operator/status.go (hunk 10 above)
Comment (sallyom) [reply pkg/operator/status.go]: yes, thanks : ).  I've kept them all as if statements (no else), instead of switch.

**Pull request merged**